pallets/treasury-buyout-extension/src/lib.rs

```diff
@@ -223,54 +222,56 @@ pub mod pallet {
 		}
 
 		/// Allows root to update the allowed currencies for buyout.
+		/// `AllowedCurrencies` storage will be reset and updated with provided `assets`.
 		///
 		/// Parameters
 		///
 		/// - `origin`: Origin must be root.
-		/// - `updates`: List of updates to be applied to `AllowedCurrencies` storage.
+		/// - `assets`: List of assets to be inserted into `AllowedCurrencies` storage.
 		///
 		/// Emits `AllowedAssetsForBuyoutUpdated` event when successful.
 		#[pallet::call_index(2)]
 		#[pallet::weight(<T as pallet::Config>::WeightInfo::update_allowed_assets())]
 		#[transactional]
 		pub fn update_allowed_assets(
 			origin: OriginFor<T>,
-			updates: Vec<BuyoutAssetUpdate<CurrencyIdOf<T>>>,
+			assets: Vec<CurrencyIdOf<T>>,
 		) -> DispatchResultWithPostInfo {
 			ensure_root(origin)?;
 
-			// Ensure number of `AllowedCurrencies` storage updates doesn't exceed the maximum allowed
+			// Ensure number of currencies doesn't exceed the maximum allowed
+			let max_allowed_currencies_for_buyout = T::MaxAllowedBuyoutCurrencies::get();
 			ensure!(
-				updates.len() <= T::MaxAllowedCurrencyUpdates::get() as usize,
-				Error::<T>::ExceedsNumberOfAllowedUpdates
+				assets.len() <= max_allowed_currencies_for_buyout as usize,
+				Error::<T>::ExceedsNumberOfAllowedCurrencies
 			);
 
-			// Used for ensuring that native token is not allowed for buyout
+			// Ensure that native token is not allowed for buyout
 			let basic_asset = <T as orml_currencies::Config>::GetNativeCurrencyId::get();
+			ensure!(
+				!assets.iter().any(|asset| *asset == basic_asset),
+				Error::<T>::NativeTokenNotAllowed
+			);
 
-			// Used for the event data
-			let mut added_assets = Vec::new();
-			let mut removed_assets = Vec::new();
-
-			for update in updates.clone() {
-				match update {
-					BuyoutAssetUpdate::Add(asset) => {
-						// Ensure native token is not added to the allowed assets for buyout
-						ensure!(asset != basic_asset, Error::<T>::NativeTokenNotAllowed);
-						AllowedCurrencies::<T>::insert(asset, ());
-						added_assets.push(Some(asset));
-					},
-					BuyoutAssetUpdate::Remove(asset) => {
-						AllowedCurrencies::<T>::remove(asset);
-						removed_assets.push(Some(asset));
-					},
+			// Clear `AllowedCurrencies` storage
+			// `AllowedCurrencies` should have at most `max_allowed_currencies_for_buyout` entries
+			let result = AllowedCurrencies::<T>::clear(max_allowed_currencies_for_buyout, None);
+			// If storage clearing returns cursor which is `Some`, then clearing was not entirely successful
```

Comment:
`Some` is returned when not all the specified number of entries in the `StorageMap` are deleted in one call. In our case, the number of entries we want to delete in one call is up to `MaxAllowedBuyoutCurrencies` which we configure in the runtime. Given that this parameter should be equal to the number of all our supported currency ids in the worst case, then I don't think we have to worry about the storage not being fully cleared in one call since it's a small number of entries after all. We can use `u32::MAX` instead of `MaxAllowedBuyoutCurrencies` as a limit so we don't have to "trust" the correctness of the runtime parameter. Honestly, I don't know how we could end up in a situation where we can't reduce the amount of allowed currencies.

Reply:
Ah true, we limit the number of items with `MaxAllowedBuyoutCurrencies` anyway. Well then we should be fine as long as we set that parameter to a reasonable number. Nevermind then.